File: src/main/scala/Awair.scala
package smick

import com.twitter.conversions.DurationOps._
import com.twitter.finagle.Http
import com.twitter.finagle.http.{ Fields, Request, RequestBuilder }
import com.twitter.util.{ TimeFormat, Future }
import java.net.URL
import scala.util.control.NonFatal

private case class AwairDevice(name: String, typ: String, id: String)
private object AwairDevice {
  def unapply(str: String): Option[AwairDevice] = {
    try {
      val Array(n, t, i) = str.split(':')
      Some(AwairDevice(n, t, i))
    } catch {
      case NonFatal(_) => None
    }
  }
}

trait Awair { self: SmickHome =>
  val awairFreq = flag("awair.freq", 5.minutes, "Loop frequency for Awair API calls")
  val awairToken = flag("awair.token", "", "Access token for Awair")
  val awairDevices = flag("awair.devices", Seq.empty[String], "List of devices: 'name:type:id'")

  private[this] val timeFormat = new TimeFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")

  def awairLoop(store: Store): Future[Unit] =
    loopIt("awair", awairFreq(), process(store))

  private def devURL(typ: String, id: String): URL =
    new URL(s"https://developer-apis.awair.is/v1/users/self/devices/${typ}/${id}/air-data/latest")

  private[this] val devices = Lazy[Seq[AwairDevice]] {
    awairDevices() collect { case AwairDevice(dev) => dev }
  }

  private[this] val clients = Lazy[Seq[(String, URL, HttpSvc, Request)]] {
    devices() map { case AwairDevice(n, t, i) =>
      val _url = devURL(t, i)
      val c = Http.client
        .withLabel(s"awair:$n")
        .withTls(_url.getHost)
        .newClient(destStr(_url))
        .toService
      val req = RequestBuilder()
        .setHeader(Fields.Authorization, s"Bearer ${awairToken()}")
        .url(_url)
        .buildGet()
      (n, _url, c, req)
    }
  }

  private def process(store: Store): Future[Unit] = Future.join {
    clients() map { case (device, url, client, req) =>
      client(req) flatMap { res =>
        val recs = json.readValue[Map[String, Any]](res.contentString)
        val data = recs("data").asInstanceOf[List[Map[String, Any]]]
        val entries = data flatMap { data =>
          val ts = {
            val raw = data("timestamp").asInstanceOf[String]
            val d = timeFormat.parse(raw)
            Some(d.inNanoseconds)
          }

          data("sensors").asInstanceOf[Seq[Map[String, Any]]] map { sensor =>
            val name = sensor("comp").asInstanceOf[String]
            StoreEntry(
              s"awair.$name", sensor("value"),
              tags = Map("device" -> device),
              time = ts)
          }
        }

        store.write(entries)
      }
    }
  }
}

File: src/internal/looper/awair.go
package looper

import (
	"context"
	"fmt"
	"github.com/rs/zerolog"
	"github.com/spf13/pflag"
	hm "github.com/sprsquish/housemetrics/internal"
	"github.com/sprsquish/housemetrics/internal/store"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type Awair struct {
	client *hm.HttpClient
	logger *zerolog.Logger

	token   string
	devices []string

	devURLs map[string]*url.URL
}

type AwairReading struct {
	Data []struct {
		Timestamp string
		Sensors   []struct {
			Comp  string
			Value interface{}
		}
	}
}

func NewAwair(name string, flags *pflag.FlagSet, logger *zerolog.Logger, client *hm.HttpClient) hm.Looper {
	a := Awair{
		client:  client,
		logger:  logger,
		devURLs: map[string]*url.URL{},
	}

	flags.StringVar(&a.token, fmt.Sprintf("%s.token", name), "", "Access token")
	flags.StringSliceVar(&a.devices, fmt.Sprintf("%s.devices", name), []string{}, "List of devices: 'name:type:id'")

	return &a
}

func (a *Awair) Init() {
	for _, device := range a.devices {
		dev := strings.Split(device, ":")
		if len(dev) != 3 {
			a.logger.Error().Str("device", device).Msg("bad device")
			continue
		}

		devName, devType, devID := dev[0], dev[1], dev[2]

		urlStr := fmt.Sprintf("https://developer-apis.awair.is/v1/users/self/devices/%s/%s/air-data/latest", devType, devID)
		if url, err := url.Parse(urlStr); err != nil {
			a.logger.Error().Err(err).Str("device", devName).Msg("couldn't parse URL")
		} else {
			a.devURLs[devName] = url
		}
	}
}

func (a *Awair) Poll(ctx context.Context, store store.Client) error {
	a.logger.Debug().Msg("polling devices")

	for devName, devURL := range a.devURLs {
		var reading AwairReading
		if err := a.client.GetJSON(ctx, &reading, func(req *http.Request) {
			req.URL = devURL
			req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", a.token))
		}); err != nil {
			return err
		}

		devTags := map[string]string{"device": devName}
		for _, entry := range reading.Data {
			ts, err := time.Parse("2006-01-02T15:04:05.999Z", entry.Timestamp)
			if err != nil {
				a.logger.Error().Err(err).Interface("reading", reading).Msg("could not parse timestamp")
			}

			for _, sensor := range entry.Sensors {
				store.Write(ctx, ts, fmt.Sprintf("awair.%s", sensor.Comp), sensor.Value, devTags)
			}
		}
	}

	return nil
}
